src/beamlime/constructors/providers.py

```diff
@@ -0,0 +1,382 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+from __future__ import annotations
+
+from functools import partial
+from typing import (
+    Callable,
+    Dict,
+    Generic,
+    ItemsView,
+    Iterator,
+    KeysView,
+    Literal,
+    Union,
+    ValuesView,
+)
+
+from beamlime.constructors.inspectors import (
+    DependencySpec,
+    Product,
+    ProductSpec,
+    ProductType,
+    UnknownType,
+    collect_argument_specs,
+    collect_attr_specs,
+    get_product_spec,
+)
+
+DependencySpecDict = Dict[str, DependencySpec]
+
+_lambda_name = (lambda: None).__qualname__
+
+
+class ProviderNotFoundError(Exception):
+    ...
+
+
+class ConflictProvidersError(Exception):
+    ...
+
+
+class ProviderExistsError(Exception):
+    ...
+
+
+class MismatchingProductTypeError(Exception):
+    ...
+
+
+def _validate_callable_as_provider(
+    callable_obj: Callable[..., Product]
+) -> Literal[True]:
+    """
+    Raises
+    ------
+    NotImplementedError
+        If the constructor of a provider is a method of a class.
+
+    Notes
+    -----
+    It is not supported to use a method of a class as a provider,
+    because it is not possible to check
+    if a ``callable_obj`` is bound to an instantiated object or just a class
+    without an access to the object containing the method.
+    Currently, it is checking if ``__qualname__`` is different from ``__name__``
+    to see if the ``callable_obj`` is a method.
+    """
+    if (
+        not isinstance(callable_obj, partial)
+        and not isinstance(callable_obj, type)
+        and hasattr(callable_obj, "__name__")
+        and not (callable_obj.__name__ == _lambda_name)
+        and hasattr(callable_obj, "__qualname__")
+        and callable_obj.__qualname__ != callable_obj.__name__
+    ):
```

Comment:
Does this not work? https://docs.python.org/3/library/inspect.html#inspect.ismethod

Reply:
Unfortunately not... it only works with the method of an instance.
And the decorator `ProviderGroup.provider` will not be called when it's instantiated,
but when it's imported,
which means `callable_obj` will be `C.foo` instead of `c.foo` in the example below.

```
provider_gr = ProviderGroup()
class C:
  @provider_gr.provider
  def foo(self):
    ...

c = C()
```

Another way is to check the type of the member of the class by `inspect.classify_class_attrs`,
and then you need to have an access to the class,
which is not possible only from the callable object.

I will maybe add a TODO tag at least to get back to this problem later
since it's not really an explicit way to check if the callable is a method....